### ISSUE TITLE
Very low priority: Possible improvement reg. `JOptionPane` in `EndRoundDelegate`.

### ISSUE THREAD

**Issue by MelvinWM** [triplea-game/triplea#6304]
Very low priority, please disregard as desired.

Reg. the comments here: https://github.com/triplea-game/triplea/blob/064d5291c00b7e54a4e1b7408f85aa927abcb237/game-core/src/main/java/games/strategy/triplea/delegate/EndRoundDelegate.java#L313-L333 , I have looked a little bit at it of interest (also because it was semi-connected to `triggerVictory` in `TriggerAttachment` and the stuff reg. `NotificationMessages` and unit testing), and I think I have an idea on how to deal with it.

I think that if `EndRoundDelegate` had an interactive step where it waits for a player to take action like in some other delegates, it might be possible to handle it that way. However, the challenge there seems to be that the player to ask might not be a standard game player (given my limited understanding of the code base), and that the current XML-defined games do not ever (as far as I know) specify any player for the "endRound" delegate. I do think that it might instead make sense to have special handling in https://github.com/triplea-game/triplea/blob/a83c5f94b893db5d97062e94480f6306d1b05683/game-core/src/main/java/games/strategy/engine/framework/ServerGame.java#L526 specific to the `EndRoundDelegate`, where it asks the right player to take action reg. whether to stop or continue the game. This is not so pretty, but it seems somewhat cleaner than the current handling, where there is also UI code.

In short:

Changes:
- Changes to `ServerGame::waitForPlayerToFinishStep()` ; specific handling of `EndRoundDelegate` with asking which player (reg. whether the game should be stopped or it should continue, after the game is over with victory).
- Changes to `EndRoundDelegate`, including `delegateCurrentlyRequiresUserInput()` and `signalGameOver()` ; Change the code from in certain cases showing a UI-dialog to instead simply ask the player using the standard delegate mechanisms on whether to continue or not.
- Other?

Benefits:
- Move UI code out of game logic and make `EndRoundDelegate` cleaner.

Costs:
- Time and development.
- Case-specific handling in `ServerGame`.
- Determining/defining the given player to ask, since it might be atypical/other.
- Other?

**Comment by DanVanAtta** [comment]
Sorry for lack of response @MelvinWM ; is this something you're still interested in pursuing?

**Comment by MelvinWM** [comment]
@DanVanAtta It's fully fine, you can just leave it as it is or close it if you want. You are welcome to close it if you want, it was primarily meant as a possible idea for approach for you to pursue in case you hadn't already considered that approach.

A part of why I looked at it is that it seemed related to some potential refactoring reg. `TriggerAttachment` (namely reg. the stuff related to unit testing, dependencies, etc. in `TriggerAttachment` and so on). Another part is that I have wondered about an approach where the game logic (GL) and the UI-related (UI) parts of the code are more separated. That seemed useful, since my thinking is that if you want to do certain kinds of stuff with game logic & state, such as state-space searching with the given state (which I believe you do in various ways such as reg. the AI as well as battle calculations), then keeping the GL and the UI code separate might well be handy and possibly worth the other trade-offs included in separating the two (especially for a game engine that is turn-based) despite it being a game engine and game logic and UI often being closely connected for game engines (at least real-time ones). I wondered whether some sort of approach such as "events" might be useful there. The role of not executing the UI code when running headless (such as (I assume) when running state-space search) currently seems to be handled to some degree by the `IDelegateBridge` abstraction, which has its advantages and drawbacks. The approach I wondered about is using events, such that when running headless, only game-logic event processing would run, and when running headed, both game-logic event processing and also UI event processing would run. However, I am uncertain about the drawbacks of such an approach (apart from the probably far-too-large amount of development work it would take to change to it, unless maybe it could be done incrementally, but even then) and how to design and implement it in a good way, .and whether it in practice would work well for this case.

That approach reg. events would likely also be hindered by the current lack of pattern matching and tagged unions in Java, though Brian Goetz and other Java language developers seem to be making good headway reg. that (though I have not looked into it greatly). See for instance "Pattern matching" and "Records" in https://en.wikipedia.org/wiki/Java_version_history#Java_SE_14 . Though, upgrading to Java 14 for this project is probably a fair distance away (it also feels a bit funny to see such large Java version numbers, though the recent version numbers of Java represents much smaller changes than earlier in Java's history).

**Comment by DanVanAtta** [comment]
> Another part is that I have wondered about an approach where the game logic (GL) and the UI-related (UI) parts of the code are more separated

> Though, upgrading to Java 14 for this project is probably a fair distance away

FWIW we've agreed to stick to the LTS versions of java.

**Comment by DanVanAtta** [comment]
> Another part is that I have wondered about an approach where the game logic (GL) and the UI-related (UI) parts of the code are more separated


100% It's an active goal to create more separation of responsibilities. Largely the code is an older coding style and of varying levels of skill. I think a lot of it suffers from under-design, where features were consistently just added, and added, and then it's difficult to untangle and features are then, well, just added because it's too costly to untangle properly.  #BBOM
